### PR TITLE
replace `trec` dataset and migrate `training_batch_hard_trec.py` from v2 to v3

### PR DESCRIPTION
**Pull request by omkar-334**
Previous  
..
Screenshot 2026-01-28 at 7 41 09 PM

New  
..
Screenshot 2026-01-28 at 7 40 56 PM

#3621

### CODE REVIEW

File: examples/sentence_transformer/training/other/training_batch_hard_trec.py
Comment (tomaarsen) [inline examples/sentence_transformer/training/other/training_batch_hard_trec.py:1]: I'm okay with not updating the download for this one - I don't have a nice TREC dataset on the Hub. If there are any nice ones, then feel free, though!

File: examples/sentence_transformer/training/other/training_batch_hard_trec.py [hunk 1]
```diff
@@ -157,23 +160,38 @@ def triplets_from_labeled_dataset(input_examples):
 logging.info("Performance before fine-tuning:")
 dev_evaluator(model)
 
-warmup_steps = int(len(train_dataloader) * num_epochs * 0.1)  # 10% of train data
+# 10% of train data
+num_train_samples = len(train_dataset)
+steps_per_epoch = num_train_samples // train_batch_size
+total_steps = steps_per_epoch * num_epochs
+warmup_steps = int(total_steps * 0.1)
```
Comment (tomaarsen) [inline examples/sentence_transformer/training/other/training_batch_hard_trec.py]: Let's go with `warmup_steps=0.1` instead

File: examples/sentence_transformer/training/other/training_batch_hard_trec.py [hunk 2]
```diff
@@ -157,23 +160,38 @@ def triplets_from_labeled_dataset(input_examples):
...
+
+# Prepare the training arguments
+args = SentenceTransformerTrainingArguments(
+    output_dir=output_path,
+    num_train_epochs=num_epochs,
+    per_device_train_batch_size=train_batch_size,
+    warmup_steps=warmup_steps,
+    # Use GROUP_BY_LABEL batch sampler for triplet losses that require multiple samples per label
+    batch_sampler=BatchSamplers.GROUP_BY_LABEL,
```
Comment (tomaarsen) [inline examples/sentence_transformer/training/other/training_batch_hard_trec.py:137]: Nice!

File: examples/sentence_transformer/training/other/training_batch_hard_trec.py [hunk 3]
```diff
@@ -157,23 +160,38 @@ def triplets_from_labeled_dataset(input_examples):
...
+trainer.train()
 
-##############################################################################
-#
 # Load the stored model and evaluate its performance on TREC dataset
-#
-##############################################################################
 
 logging.info("Evaluating model on test set")
 test_evaluator = TripletEvaluator.from_input_examples(test_set, name="trec-test")
```
Comment (tomaarsen) [inline examples/sentence_transformer/training/other/training_batch_hard_trec.py]: Can we save after training? E.g. https://github.com/huggingface/sentence-transformers/pull/3631#discussion_r2740927619

File: examples/sentence_transformer/training/other/training_batch_hard_trec.py [hunk 4]
```diff
@@ -157,23 +160,38 @@ def triplets_from_labeled_dataset(input_examples):
...
+    num_train_epochs=num_epochs,
+    per_device_train_batch_size=train_batch_size,
+    warmup_steps=warmup_steps,
+    # Use GROUP_BY_LABEL batch sampler for triplet losses that require multiple samples per label
+    batch_sampler=BatchSamplers.GROUP_BY_LABEL,
+    eval_strategy="steps",
+    eval_steps=1000,
+    save_strategy="steps",
+    save_steps=1000,
+    logging_steps=100,
```
Comment (tomaarsen) [inline examples/sentence_transformer/training/other/training_batch_hard_trec.py]: Can we use ratios here? E.g. eval_steps=0.1, logging_steps=0.01?

File: examples/sentence_transformer/training/other/training_batch_hard_trec.py
Comment (omkar-334) [reply examples/sentence_transformer/training/other/training_batch_hard_trec.py:1]: Hey, there's a dataset at https://huggingface.co/datasets/CogComp/trec, although it uses a loading script... Maybe you can update that one or just upload a new one.   
The number of samples in both train and test splits are matching. However, the labels seem to be different. Can you check it out and let me know?

Comment (omkar-334) [reply examples/sentence_transformer/training/other/training_batch_hard_trec.py:1]: this one is uploaded by the original authors themselves, the CogComp group

Comment (omkar-334) [reply examples/sentence_transformer/training/other/training_batch_hard_trec.py:1]: image  

An image for quick reference

Comment (omkar-334) [reply examples/sentence_transformer/training/other/training_batch_hard_trec.py:1]: I can maybe make a lookup table and assign the current `label` and `guid` and reupload to hf?

Comment (tomaarsen) [reply examples/sentence_transformer/training/other/training_batch_hard_trec.py:1]: There's also an issue on the labels: https://huggingface.co/datasets/CogComp/trec/discussions/1 Although I'm still a bit confused by it.

Feel free to reupload a fixed version without a loading script, if you want. We can then either keep it on your user or move it to the `sentence-transformers` org, depending on your preference.

Comment (omkar-334) [reply examples/sentence_transformer/training/other/training_batch_hard_trec.py:1]: I've uploaded a fixed version to my own user.. I inspected the labels issue and turns out  - 
1. The issue on the labels - there were 47 labels before, they fixed it to contain 50 labels
2. Our case - The original dataset link in this code has 50 labels, the hf dataset has 50 labels. I checked the mapping and it is consistent (strict one-to-one mapping). So it means that only the label indexes have changed. This is trivial and we can proceed with the hf dataset. (the issue is irrelevant)

**Comment by omkar-334** [comment]
Comparing performance  
Old Dataset - 
Screenshot 2026-01-31 at 4 15 47 PM  

New Dataset - 
Screenshot 2026-01-31 at 4 16 06 PM

**Comment by omkar-334** [comment]
hey @tomaarsen thanks for the help on this!  
this PR is ready to be merged. Let me know if there is anything else to be changed.

**Comment by omkar-334** [comment]
2 test failing, the `3.13` one seems to be because of `429 Too Many Requests`
